### PR TITLE
[HIP] Update to 4.2.0

### PR DESCRIPTION
**Pull request by jpsamaroo**
I still need to test if this can build dependent libraries like rocBLAS, but let's make sure this passes CI.

### CODE REVIEW

**Comment by jpsamaroo** [comment]
I ran into the same issue as ROCmOpenCLRuntime, and tried fixing it with `patchelf --set-rpath '$ORIGIN/../lib64:$ORIGIN/../lib:$ORIGIN:$ORIGIN/../hsa/lib' "${prefix}/hip/lib/libamdhip64.so"`, but this doesn't seem to help. @giordano any recommendations?

**Comment by giordano** [comment]
I don't think it's the same issue. I see there is a file called `libamdhip64.so.4.2.21516-`, is that a [valid file name](https://github.com/JuliaPackaging/BinaryBuilderBase.jl/blob/030c6e47d0e2bb9f1b146b9e5a8d3231e9c437b5/src/Products.jl#L204)? I'm away from computer right now and I can't check myself, but I guess the dash breaks the regex-matching.

**Comment by jpsamaroo** [comment]
@giordano still not working. Also, it seems like BB is trying to add a symlink that already exists; is this a bug?

**Comment by giordano** [comment]
I'm looking into hsa-rocr insane symlinks, I think that's the root of all of our problems

File: H/HIP/build_tarballs.jl [hunk 1]
```diff
@@ -20,38 +19,47 @@ cd ${WORKSPACE}/srcdir/HIP*/
...
 # link to .hipVersion so clang can find it
 ln -s ../hip/bin/.hipVersion ${bindir}/.hipVersion
+
+# Manually patch runpaths of `libamdhip64.so`, but we need to investigate why
+# this is happening.  For more information see
+# https://github.com/JuliaPackaging/Yggdrasil/pull/4077#issuecomment-999926020
+if [[ "${target}" == *-linux-gnu ]]; then
+    echo "Patching RPATH"
+    patchelf --set-rpath '$ORIGIN/../lib64:$ORIGIN/../lib:$ORIGIN:$ORIGIN/../hsa/lib' "${prefix}/hip/lib/libamdhip64.so"
+fi
```
Comment (giordano) [inline H/HIP/build_tarballs.jl:54]: This isn't needed anymore with #4177
```suggestion
```

File: H/HIP/build_tarballs.jl [hunk 2]
```diff
@@ -20,38 +19,47 @@ cd ${WORKSPACE}/srcdir/HIP*/
...
-
-# force hipcc to pass paths to clang
-atomic_patch -p1 "${WORKSPACE}/srcdir/patches/hipcc-force-paths.patch"
-
 apk add coreutils dateutils
 
 mkdir build && cd build
+ln -s ${prefix}/bin/clang ${prefix}/tools/clang
 export HIP_CLANG_PATH=${prefix}/tools
 cmake -DCMAKE_INSTALL_PREFIX=${prefix}/hip \
```
Comment (giordano) [inline H/HIP/build_tarballs.jl:27]: I don't think it's the problem here, but why we don't use simply `${prefix}`?

**Comment by giordano** [comment]
The symlink we create comes from https://github.com/JuliaPackaging/BinaryBuilder.jl/blob/f232f87d9928fcefbb46ec775c62cac0c3a07897/src/auditor/soname_matching.jl#L101-L133.  It fails because there are broken links:
```console
% ls -l
total 2896
drwxr-xr-x 4 mose mose    4096 Jan  5 23:13 cmake
-rw-r--r-- 1 mose mose  269516 Jan  5 23:13 hip_prof_str.h
lrwxrwxrwx 1 mose mose      16 Jan  5 23:13 libamdhip64.so -> libamdhip64.so.4
lrwxrwxrwx 1 mose mose      25 Jan  5 23:13 libamdhip64.so.4 -> libamdhip64.so.4.2.22013-
-rwxr-xr-x 1 mose mose 2683240 Jan  5 23:14 libamdhip64.so.4.2.22013
lrwxrwxrwx 1 mose mose      85 Jan  5 23:13 libamdhip64.so.4.2.22013- -> ../../../x86_64-linux-gnu-libgfortran5-cxx11/destdir/hip/lib/libamdhip64.so.4.2.22013
```

**Comment by giordano** [comment]
It may be easier to fix the build system rather than fiddling with symlinks that bitrot

**Comment by giordano** [comment]
Wait, this just works now? :open_mouth:

**Comment by jpsamaroo** [comment]
Apparently so! I still need some time to test it before we merge, though.

**Comment by jpsamaroo** [comment]
Ready to go! I might need some help getting downstream libs to cooperate on musl, but at least rocRAND 4.2.0 builds successfully with this.

**Pull request merged**